---
title: "BIOS 201 Workshop 3"
date: "3/18/2020"
output: html_document
---

# Workshop 3: Differential expression analysis of RNA-seq data

## Introduction

Last week, we mapped RNA-seq reads from a [study](http://journals.plos.org/plosone/article/file?type=supplementary&id=info:doi/10.1371/journal.pone.0097550.s002) comparing lung tissues taken from diseased and healthy individuals. This week, weâ€™ll pick up the analysis where we left off. 

We will identify genes differentially expressed between IPF and normal lung tissues. Broadly speaking, this might allow us to 1) come up with diagnostic genetic screens for IPF susceptibility or incidence, or 2) infer possible mechanisms of the disease for further follow-up.

We'll use the R package _DESeq2_ to identify differentially expressed genes.

## Goals

This workshop will enable you to do the following:

* Compare expression across samples
* Perform principal component analysis (PCA)
* Remove the effects of known experimental confounders
* Identify differentially expressed genes with DESeq2
* Verify and follow up DESeq2 results

## Setup

We'll be doing today's workshop entirely in RStudio; however, we first need to copy a few data files from rice to your local machine. Using Command Prompt in Windows or Terminal in Mac, `cd` (e.g. `cd ~/Documents/BIOS201`) to the folder in which you want to copy the data files. You should copy the whole workshop3 folder from rice:

On Mac:
```
scp -r <yourname>@rice.stanford.edu:/afs/ir/class/bios201/workshop2/workshop3 .
```

Or on Windows:
```
pscp -r <yourname>@rice.stanford.edu:/afs/ir/class/bios201/workshop2/workshop3 .
```

**Note:** You can do this in the `Terminal` tab in RStudio, at least for Mac! (Not 100% sure about Windows)  

We are going to use this R Markdown (.Rmd) file to walk through today's workshop. Each `Rmd` file has an initial `setup` chunk that applies some settings to all of the code chunks.  

```{r setup, echo = F}
# when you "knit" this file, do you want the resulting PDF to print the code in each chunk (TRUE = yes)?
knitr::opts_chunk$set(echo = TRUE)

################################################################################
# set your working directory (wherever you downloaded Workshop3 files)
#### YOU NEED TO CHANGE THIS PATH TO THE APPROPRIATE PATH
knitr::opts_knit$set(root.dir = '~/OneDrive/BIOS201/workshop3/')
################################################################################
# note that outside of an Rmd code chunk, use `setwd()` to set the working directory in R
```

As in a regular R script in RStudio, you can run a single line of code with Command-Enter (Mac) or Ctrl-Enter (PC). You can run a whole chunk of code with Command(/Ctrl) + Shift + Enter **or** click the little ">" button in the top-right corner of the chunk.

To comment or uncomment a bunch of lines, highlight the lines and use Command(/Ctrl) + Shift + C.  

First, you need to install two R packages, DESeq2 and pheatmap. If you completed last week's workshop, you'll already have both installed. If not, you can install them using the following code:  
```{r install packages, eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
install.packages("pheatmap")
```

Then load the packages.
```{r load libraries}
library(DESeq2)
library(pheatmap)
```

## Exploratory Data Analysis

Before we start testing for differential expression, it's always important to first visualize our data and perform quality control checks. We'll start by looking at `counts.txt`, which contains the log-transformed IPF gene expression from last week with some simulated sequencing batch effects.

Read in the matrix of log-transformed gene expression counts. Verify using the `head` command that your counts matrix loaded properly.

```{r read counts}
log_counts = read.table("counts.txt")
head(log_counts)
```

As you saw at the end of the last workshop, one of our samples, "Norm7", was not actually from lung tissue; this sample doesn't express the lung surfactant proteins. We've already removed this sample from the analysis, so you won't have to worry about it.  

First, check the size of your data using `dim(log_counts)`.  
```{r q genes}
dim(log_counts)
```

:question: How many genes are measured in this dataset?  

If there's a significant difference between gene expression in healthy and diseased individuals, then we should see that samples of similar types are more correlated with one another than with samples of different types. We can test whether this is the case by making a scatterplot in R.
```{r scatterplot}
# Compare counts in two normal samples
plot(log_counts[,1], log_counts[,3])
abline(0,1)

# # you can do this with ggplot too:
# library(ggplot2)
# ggplot(log_counts, aes(x=Norm1, y=Norm3)) +
#   geom_point() +
#   geom_abline() +
#   theme_classic()
```

Now compute the correlation of gene expression between these two samples.
```{r correlation}
cor(log_counts[,1], log_counts[,3])
```

Now, modify the previous code to make a scatterplot comparing the expression in a healthy sample and a diseased sample. (Remember, you can select the _n_-th column of your `log_counts` matrix by typing `log_counts[,n]`).
```{r healthy v diseased}
head(log_counts)
plot(log_counts[,'Norm1'], log_counts[,'IPF1'])
# or plot(log_counts[,1], log_counts[,8])
abline(0,1)
cor(log_counts[,'Norm1'], log_counts[,'IPF1'])
```

:question: Which pair of samples has greater correlation?

We can also compute the pairwise correlation between all the different samples at once:
```{r all correlations}
cor(log_counts)
```

Now view this matrix as a heatmap:
```{r heatmap}
log_cor <- cor(log_counts)
diag(log_cor) <- NA
pheatmap(log_cor, show_rownames = TRUE)
```

:question: Do the healthy and diseased individuals separate into distinct clusters?

Even if the healthy and diseases samples don't separate into perfect clusters, that's not always a problem. But plotting heatmaps can help us to identify batch effects in our data.

## Detecting batch effects

As we noticed in the last section, some of the similar samples cluster together, but we don't observe two well-defined groups of diseased and healthy samples. In this case, it's because we've spiked in a simulated sequencing batch effect, which causes some of the samples from the same sequencing run to appear more similar than they really are.

[Batch effects](http://www.molmine.com/magma/global_analysis/batch_effect.html) are unwanted patterns in sequencing data that arise because of the way the samples have been handled before or during sequencing. For example, if some of the samples were left on the bench overnight but others were left in the freezer, then we might observe serious batch effects. Similarly, if samples were analyzed in different sequencing runs (or even different lanes within the same sequencing run) batch effects may be visible.

If we're not careful, batch effects can completely overpower the signal we're interested in observing. In some of the worst cases, researchers have mistaken batch effects for biological signal and didn't notice this problem until long after publication. Luckily, there are several precautions we can take to minimize the risk of false results due to batch effects.

### Loading covariates

It's always a good idea to note which samples were sequenced on which sequencing run, as well as other important metadata. We've supplied a file `covariates.txt` that contains some such information. Load it into R:
```{r load covariates}
covariates = read.table("covariates.txt")
```

Take a look at the covariates.
```{r}
head(covariates)
table(covariates$sex)
```

:question: How many IPF females are included in this dataset?

### Principal Component Analysis

A method called [Principal Component Analysis](https://en.wikipedia.org/wiki/Principal_component_analysis), or PCA, is used to detect and visualize batch effects in sequencing data. When batch effects or other effects due to experimental covariates are present in the data, they can cause major variation in many of the samples' gene expression; however, if we're lucky, the same set of genes will be affected similarly across all the affected samples. (For example, if you leave your samples out on the bench overnight, you might expect that certain RNAs will degrade more rapidly than other long-lasting RNAs, so the former RNAs will appear underexpressed in all of the affected samples.)

PCA is an algorithm that identifies sets of genes that vary similarly across samples. At a high level, PCA finds sources of variation across samples, and then outputs values called _principal components_ that represent the sets of varying genes and to what extent they are present in each sample. If batch effects are present, they will often be detectable by looking at the first few principal components of the gene expression matrix.

Let's try PCA on our RNA-seq data. We can visualize the first two principal components as follows. Our points represent samples, and are color-coded by disease status.

```{r pca color by condition}
pca = prcomp(t(log_counts)) # "features"/measurements (genes in our case) need to be in the columns

# A simple function for color-coding points
color_code <- function(x, vals, colors)
{
  i = which(x == vals)
  return(colors[i])
}

# Plot PC1 and PC2
color_status = sapply(substring(covariates$samples, 1, 3), FUN=color_code,
                      vals=c("IPF", "Nor"), colors=c("blue", "red"))
plot(pca$x[,1], pca$x[,2], col=color_status, pch=16)

# # alternatively, use ggplot:
# head(pca$x)
# pcs = data.frame(pca$x)
# pcs$status = as.character(covariates$status[match(rownames(pca$x), as.character(covariates$samples))])
# ggplot(pcs, aes(x=PC1, y=PC2, color=status)) +
#   geom_point() +
#   theme_classic() +
#   scale_colour_manual(values=c(Norm='red',IPF='blue'))
```

We can also plot the samples color-coded by sequencing batch:
```{r color by batch}
color_seqbatch = sapply(covariates$seq.batch, FUN=color_code,
                        vals=c(1, 2), colors=c("blue", "red"))
plot(pca$x[,1], pca$x[,2], col=color_seqbatch, pch=16)
```

:question: Which variable (disease status or sequencing batch) causes a stronger clustering effect in our data?

:question: Which of the principal components (#1 or #2) separates the samples by this variable?

As you can see, there's a clear batch effect in our data. However, we can still attempt to explore the disease's influence on expression by simply ignoring the effects of these first few principal components. Check out PC3 and PC4 instead.
```{r later PCs}
# Plot PC3 and PC4
plot(pca$x[,3], pca$x[,4], col=color_status, pch=16)
```

:question: Now do you see any separation of points based on disease status?

If we plot the correlation heatmap after removing the first 2 principal components, we now see that the samples cluster more nicely by disease status:
```{r remove batch}
pca_mat = cor(t(as.matrix(pca$x[,3:ncol(pca$x)])))
diag(pca_mat) = NA
pheatmap(pca_mat)
```

Removing batch effects is an important first step in any gene expression analysis. It can be particularly tricky when the experimental covariates are not recorded; however, software tools such as [PEER](https://www.ncbi.nlm.nih.gov/pubmed/22343431) can be used to detect unknown covariates. **Note:** PEER is incompatible with the newest version of R (3.6). You may want to investigate alternative methods. 

## Differential expression

Now that we've eliminated batch effects, we're ready to test for differentially expressed genes. We achieve this using an R tool called [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) that quantifies the extent to which gene expression changes between conditions.

As input, DESeq2 takes non-normalized counts values. DESeq2 statistically models the probability of differential expression, which involves placing greater confidence on differentially expressed genes for which a greater number of counts are observed across samples. To simplify this part of the analysis, we've supplied the original read count data from last week's exercise with simulated sequencing batch effects removed. It is possible that other sequencing batch effects still remain since we've never explicitly corrected these. Normally we should attempt to detect these batch effects using a tool such as PEER for identifying "hidden" effects; for today's workshop, we'll ignore this aspect of the analysis.

Load the denoised expression counts from `counts_denoised.txt` and use these counts for the remainder of the workshop.

```{r adjusted counts}
counts = read.table("counts_denoised.txt")
```

Run DESeq using the following commands:
```{r deseq}
rownames(covariates) = covariates$samples
head(covariates)
dds <- DESeqDataSetFromMatrix(countData <- counts,
                              colData <- covariates,
                              design = ~ factor(status, levels=c("Norm","IPF")))
dds <- DESeq(dds, betaPrior=FALSE)
res_nosex = results(dds) # this gets the differentially expressed genes for a contrast
## because we set the levels in the design contrast, this is equivalent to:
# results(dds, contrast=c('status','IPF','Norm'))
```

DESeq produces an _MA plot_ that shows the estimated log fold change in gene expression in diseased individuals relative to healthy individuals, plotted against the mean expression of the gene across samples. Points plotted in red are significantly differentially expressed between samples. 
```{r MA plot}
plotMA(dds)
```

:question: What is the relationship between the mean expression and the magnitude of log fold change?

:question: Do you see more highly overexpressed or underexpressed genes?

We can also directly correct for known covariates using DESeq2. Here we correct any effects due to the individual's sex. Note that we've added an additional term to the design formula.
```{r adjust for sex}
dds <- DESeqDataSetFromMatrix(countData <- counts,
                              colData <- covariates,
                              design = ~ sex + factor(status, levels=c("Norm","IPF")))
dds <- DESeq(dds, betaPrior=FALSE)
res = results(dds)
plotMA(dds)
```

DESeq provides two different p-values, `pvalue` and `padj`. When searching for differential expression across many genes, you should always use `padj`, which corrects for [multiple testing](http://www.stat.berkeley.edu/~mgoldman/Section0402.pdf).

Compare the strength of results between the sex-corrected DESeq2 run and the run with no covariates corrected.
```{r compare contrasts}
plot(-log10(res_nosex$padj), -log10(res$padj), xlim=c(0,30), ylim=c(0,30),
            xlab="-log10 p-value, No sex correction", ylab="-log10 p-value, sex correction")
abline(a=0,b=1)
abline(a=-log10(0.05),b=0,lty=3,col="red")
abline(v=-log10(0.05),lty=3,col="red")
```

:question: Are there any genes that become significant after correcting for sex? (Use the dashed red line as a cutoff for significance.) Do any that were initially significant become insignificant after correction?

Let's see how many genes are differentially expresssed. Again, use adjusted p-values.  

```{r n degs}
sum(res_nosex$padj < 0.05, na.rm=TRUE)
sum(res$padj < 0.05, na.rm=TRUE)
```

:question: How many genes are differentially expressed at an adjusted p-value of 0.05 in each DESeq2 run? Do we find more differentially expressed genes before or after removing the sex effect?

For the remainder of this workshop, we'll use only the results with the sex effect removed.

What are our top differentially expressed genes?

```{r order res}
head(res[order(res$padj),])
```

According to DESeq, one of our top differentially expressed genes is ENSG00000170962. Make a boxplot to verify this.
```{r single gene}
status <- c(rep("Norm", 7), rep("IPF", 8))
boxplot(as.numeric(log_counts[rownames(log_counts)=="ENSG00000170962",]) ~ status)

# # to do this with ggplot:
# ggplot(data.frame(t(log_counts)), aes(x=status,y=ENSG00000170962)) +
#   geom_boxplot() +
#   theme_classic()
```

### Following up results

ENSG00000170962 isn't a very descriptive name. Look up this gene on [GeneCards](http://www.genecards.org/cgi-bin/carddisp.pl?gene=PDGFD&keywords=ENSG00000170962) to find its common name.

There are also packages that allow us to query gene names within R. 
```{r biomart, eval=F}
# install.packages('biomaRt')
library(biomaRt)
# get h sapiens biomaRt
# this takes a minute
ensembl_mart = useMart("ensembl", dataset="hsapiens_gene_ensembl") 

# look at available fields in the database. there's a lot of info here
listAttributes(mart = ensembl_mart)  
# select the attributes you want to include in your table
# I'm just including gene IDs and gene names
# this takes a minute 
hs = getBM(attributes=c("external_gene_name",
                        "ensembl_gene_id",
                        "entrezgene_id",
                        "wikigene_description"),mart=ensembl_mart)
# filter down to genes in our analysis
hs = hs[hs$ensembl_gene_id %in% rownames(res),]
head(hs)
# look for a single gene 
hs[hs$ensembl_gene_id == "ENSG00000170962",]
hs[hs$external_gene_name == "ACTA1",]

## alternatively, if you only care about gene symbols and IDs, org.Hs.eg.db is much faster:
# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("clusterProfiler")
library(org.Hs.eg.db)
library(clusterProfiler)
gene_map = data.frame(bitr(geneID = rownames(res), 
                           fromType = 'ENSEMBL', 
                           toType = 'SYMBOL', 
                           OrgDb = org.Hs.eg.db, drop = TRUE))
gene_map[gene_map$ENSEMBL == "ENSG00000170962",]
gene_map[gene_map$SYMBOL == "ACTA1",]
```

[Previous work](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0092111#) has implicated MUC5B and DSP as differentially expressed genes in IPF. Test these genes to see if they're replicated in our analysis.
```{r MUC5B}
# MUC5B
res[rownames(res) == "ENSG00000117983",]$padj
```

Try the next one on your own. Look up the ENSG identifier for DSP online (or use `biomaRt` or `org.Hs.eg.db` as above), and then get that row from the DESeq2 results.
```{r try another}

```

:question: Which of these two genes has greater differential expression in our cohort?

To view the original code used by Tracy Nance for this analysis, see
https://github.com/datapixie/ipf/blob/master/runDESeq/runDESeq.R.
